contributors/design-proposals/storage/volume-topology-scheduling.md

```diff
@@ -328,30 +693,54 @@ avoid these error conditions are to:
 * Separate out volumes that the user prebinds from the volumes that are
 available for the system to choose from by StorageClass.
 
-#### PV Controller Changes
+### PV Controller Changes
 When the feature gate is enabled, the PV controller needs to skip binding
 unbound PVCs with VolumBindingWaitForFirstConsumer and no prebound PVs
 to let it come through the scheduler path.
 
 Dynamic provisioning will also be skipped if
-VolumBindingWaitForFirstConsumer is set.  The scheduler will signal to
+VolumBindingWaitForFirstConsumer is set. The scheduler will signal to
 the PV controller to start dynamic provisioning by setting the
-`annStorageProvisioner` annotation in the PVC.
+`annSelectedNode` annotation in the PVC. If provisioning fails, the PV
+controller can signal back to the scheduler to retry dynamic provisioning by
+removing the `annSelectedNode` annotation.
 
 No other state machine changes are required.  The PV controller continues to
 handle the remaining scenarios without any change.
 
 The methods to find matching PVs for a claim and prebind PVs need to be
 refactored for use by the new scheduler functions.
 
-#### Scheduler Changes
+### Dynamic Provisioning interface changes
+The dynamic provisioning interfaces will be updated to pass in:
+* selectedNode, when late binding is enabled on the StorageClass
+* allowedTopologies, when it is set in the StorageClass
+
+If selectedNode is set, the provisioner should get its appropriate topology
+labels from the Node object, and provision a volume based on those topology
+values. In the common use case for a volume supporting a single topology domain,
+if nodeName is set, then allowedTopologies can be ignored.
+
+In-tree provisioners:
+```
+Provision(selectedNode *string, allowedTopologies *storagev1.VolumeProvisioningTopology) (*v1.PersistentVolume, error)
```

Comment:
What if the entire Node spec is passed instead of just the node name? A topology-aware volume plugin has to fetch the node object anyway in order to get the node's topology info.

Alternatively - should there be a dedicated NodeTopology field inside Node spec? Then here we could just pass that field, and volume plugins encode their topology information in this field rather than labels.

A related issue: https://github.com/kubernetes/kubernetes/issues/61906 (TL;DR if we had this field, it could be passed to applications through downwardAPI, making it easy for applications to obtain topology info).

Reply:
Having a dedicated topology field in the node object could be useful, but I think it has broader implications than just for storage, so would need to be considered more holistically.

Passing a Node object sounds ok to me. It means that the PV controller also needs permissions to get Node objects.

Reply:
@msau42 As we discussed before, we need a way to calculate the capacity used by released PVs(i.e. provisioned PVs whose claims got deleted then)

so here we'll also need to pass `AnnProvisionedTopology`, and require the developers of provisioners to add the annotation onto PVs.

Reply:
done